src/txrebroadcast.cpp

```diff
@@ -0,0 +1,242 @@
+#include <script/script.h>
+#include <txrebroadcast.h>
+#include <util/time.h>
+
+#include <boost/multi_index/hashed_index.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index_container.hpp>
+
+/** We rebroadcast upto 3/4 of max block weight to reduce noise due to
+ * circumstances such as miners mining priority transactions. */
+static constexpr unsigned int MAX_REBROADCAST_WEIGHT{3 * MAX_BLOCK_WEIGHT / 4};
+
+/** Default minimum age for a transaction to be rebroadcast */
+static constexpr std::chrono::minutes REBROADCAST_MIN_TX_AGE{30min};
+
+/** Maximum number of times we will rebroadcast a transaction */
+static constexpr int MAX_REBROADCAST_COUNT{6};
+
+/** Minimum amount of time between returning the same transaction for
+ * rebroadcast */
+static constexpr std::chrono::hours MIN_REATTEMPT_INTERVAL{4h};
+
+/** The maximum number of entries permitted in m_attempt_tracker */
+static constexpr int MAX_ENTRIES{500};
+
+/** The maximum age of an entry ~3 months */
+static constexpr std::chrono::hours MAX_ENTRY_AGE{24h * 30 * 3};
+
+struct RebroadcastEntry {
+    RebroadcastEntry(std::chrono::microseconds now_time, uint256 wtxid)
+        : m_last_attempt(now_time),
+          m_wtxid(wtxid) {}
+
+    std::chrono::microseconds m_last_attempt;
+    const uint256 m_wtxid;
+    int m_count{1};
+};
+
+/** Used for multi_index tag  */
+struct index_by_last_attempt {};
+
+class indexed_rebroadcast_set : public
+boost::multi_index_container<
+    RebroadcastEntry,
+    boost::multi_index::indexed_by<
+        // sorted by wtxid
+        boost::multi_index::hashed_unique<
+            boost::multi_index::tag<index_by_wtxid>,
+            boost::multi_index::member<RebroadcastEntry, const uint256, &RebroadcastEntry::m_wtxid>,
+            SaltedTxidHasher
+        >,
+        // sorted by last rebroadcast time
+        boost::multi_index::ordered_non_unique<
+            boost::multi_index::tag<index_by_last_attempt>,
+            boost::multi_index::member<RebroadcastEntry, std::chrono::microseconds, &RebroadcastEntry::m_last_attempt>
+        >
+    >
+>{};
+
+TxRebroadcastHandler::~TxRebroadcastHandler() = default;
+
+TxRebroadcastHandler::TxRebroadcastHandler(const CTxMemPool& mempool, const ChainstateManager& chainman, const CChainParams& chainparams)
+    : m_mempool{mempool},
+      m_chainman{chainman},
+      m_chainparams(chainparams),
+      m_attempt_tracker{std::make_unique<indexed_rebroadcast_set>()}{}
+
+std::vector<TxIds> TxRebroadcastHandler::GetRebroadcastTransactions(const CBlockIndex* recent_block_index)
+{
+    std::vector<TxIds> rebroadcast_txs;
+    auto start_time = GetTime<std::chrono::microseconds>();
+
+    int rebroadcast_block_weight = MAX_REBROADCAST_WEIGHT;
+    CBlock block;
+    const Consensus::Params& consensus_params = m_chainparams.GetConsensus();
+    if (ReadBlockFromDisk(block, recent_block_index, consensus_params)) {
```

Comment:
yeah, I went through a similar thought process. In the initial pass, I opted for the `ReadBlockFromDisk` approach because it seemed simpler. But in the latest push, I updated the `UpdateBlockTip` function signature to also take in `block`. Unfortunately, it seems like callers cannot rely on `block` to be properly populated, so I still need to have the `ReadBlockFromDisk` as a fallback. The relevant code is in 33eb966d284bd954af16ed5905315e0bc91c3662 & 33eb966d284bd954af16ed5905315e0bc91c3662, let me know what you think.